python/ray/worker.py

```diff
@@ -2376,9 +2403,26 @@ def wait(object_ids, num_returns=1, timeout=None, worker=global_worker):
         if num_returns > len(object_ids):
             raise Exception("num_returns cannot be greater than the number "
                             "of objects provided to ray.wait.")
+
+        # Get the task ID, to notify the backend which task is blocked.
+        with worker.state_lock:
+            current_task_id = worker.current_task_id
+            if not ray.utils.is_main_thread():
+                # If this is running on a separate thread, then the mapping
+                # to the current task ID may not be correct. Generate a
+                # random task ID so that the backend can differentiate
+                # between different threads.
+                task_id = ray.ObjectID(random_string())
+                if not worker.multithreading_warned:
+                    logger.warning(
+                        "Calling ray.get or ray.wait in a separate thread may "
+                        "lead to deadlock")
+                    worker.multithreading_warned = True
+            assert task_id.id() != NIL_ID
```

Comment:
Also, this should be `current_task_id.id()` right? (and the above assign should be to current_task_id).